core/vm/contracts_lightclient.go

```diff
@@ -397,3 +401,72 @@ type cometBFTLightBlockValidateHertz struct {
 func (c *cometBFTLightBlockValidateHertz) Run(input []byte) (result []byte, err error) {
 	return c.run(input, true)
 }
+
+// tmSignatureRecover implemented as a native contract.
+type tmSignatureRecover struct{}
+
+func (c *tmSignatureRecover) RequiredGas(input []byte) uint64 {
+	return params.EcrecoverGas
+}
+
+const (
+	tmPubKeyLength       uint8 = 33
+	tmSignatureLength    uint8 = 64
+	tmSignatureMsgLength uint8 = 32
+)
+
+func (c *tmSignatureRecover) Run(input []byte) (result []byte, err error) {
+	if len(input) != int(tmPubKeyLength)+int(tmSignatureLength)+int(tmSignatureMsgLength) {
+		return nil, fmt.Errorf("invalid input")
+	}
+
+	return c.runTMSecp256k1Signature(
+		input[:tmPubKeyLength],
+		input[tmPubKeyLength:tmPubKeyLength+tmSignatureLength],
+		input[tmPubKeyLength+tmSignatureLength:],
+	)
+}
+
+func (c *tmSignatureRecover) runTMSecp256k1Signature(pubkey, signatureStr, msgHash []byte) (result []byte, err error) {
```

Comment:
Please reuse most of the code here: https://github.com/bnb-chain/bnc-tendermint/blob/982be8c723615a1d75fb43170565f71e058734a0/crypto/secp256k1/secp256k1.go#L231

Reply:
done